zebra-chain/src/types/amount.rs

```diff
@@ -0,0 +1,387 @@
+
+        Ok(())
+    }
+
+    #[test]
+    fn test_add_opt_rhs() -> Result<()> {
+        zebra_test::init();
+        let one: Amount = 1.try_into()?;
+        let neg_one: Amount = (-1).try_into()?;
+        let neg_one = Ok(neg_one);
+
+        let zero: Amount = 0.try_into()?;
+        let new_zero = one + neg_one;
+
+        assert_eq!(zero, new_zero?);
+
+        Ok(())
+    }
+
+    #[test]
+    fn test_add_opt_both() -> Result<()> {
+        zebra_test::init();
+        let one: Amount = 1.try_into()?;
+        let one = Ok(one);
+        let neg_one: Amount = (-1).try_into()?;
+        let neg_one = Ok(neg_one);
+
+        let zero: Amount = 0.try_into()?;
+        let new_zero = one.and_then(|one| one + neg_one);
+
+        assert_eq!(zero, new_zero?);
+
+        Ok(())
+    }
+
+    #[test]
+    fn test_add_assign() -> Result<()> {
+        zebra_test::init();
+        let one: Amount = 1.try_into()?;
+        let neg_one: Amount = (-1).try_into()?;
+        let mut neg_one = Ok(neg_one);
+
+        let zero: Amount = 0.try_into()?;
+        neg_one += one;
+        let new_zero = neg_one;
+
+        assert_eq!(Ok(zero), new_zero);
+
+        Ok(())
+    }
+
+    #[test]
+    fn test_sub_bare() -> Result<()> {
+        zebra_test::init();
+        let one: Amount = 1.try_into()?;
+        let zero: Amount = 0.try_into()?;
+
+        let neg_one: Amount = (-1).try_into()?;
+        let new_neg_one = zero - one;
+
+        assert_eq!(Ok(neg_one), new_neg_one);
+
+        Ok(())
+    }
+
+    #[test]
+    fn test_sub_opt_lhs() -> Result<()> {
+        zebra_test::init();
+        let one: Amount = 1.try_into()?;
+        let one = Ok(one);
+        let zero: Amount = 0.try_into()?;
+
+        let neg_one: Amount = (-1).try_into()?;
+        let new_neg_one = zero - one;
+
+        assert_eq!(Ok(neg_one), new_neg_one);
+
+        Ok(())
+    }
+
+    #[test]
+    fn test_sub_opt_rhs() -> Result<()> {
+        zebra_test::init();
+        let one: Amount = 1.try_into()?;
+        let zero: Amount = 0.try_into()?;
+        let zero = Ok(zero);
+
+        let neg_one: Amount = (-1).try_into()?;
+        let new_neg_one = zero - one;
+
+        assert_eq!(Ok(neg_one), new_neg_one);
+
+        Ok(())
+    }
+
+    #[test]
+    fn test_sub_assign() -> Result<()> {
+        zebra_test::init();
+        let one: Amount = 1.try_into()?;
+        let zero: Amount = 0.try_into()?;
+        let mut zero = Ok(zero);
+
+        let neg_one: Amount = (-1).try_into()?;
+        zero -= one;
+        let new_neg_one = zero;
+
+        assert_eq!(Ok(neg_one), new_neg_one);
+
+        Ok(())
+    }
+
+    #[test]
+    fn add_with_diff_constraints() -> Result<()> {
+        let one = Amount::<NonNegative>::try_from(1)?;
+        let zero = Amount::<NegativeAllowed>::try_from(0)?;
+
+        (zero - one.constrain()).expect("should allow negative");
+        (zero.constrain() - one).expect_err("shouldn't allow negative");
+
+        Ok(())
+    }
```

Comment:
@hdevalence new test showing the new constrain API for adding Amount's with different constraints

Reply:
this is really slick